
package net.mcreator.deaceased.entity;

import software.bernie.geckolib.util.GeckoLibUtil;
import software.bernie.geckolib.core.object.PlayState;
import software.bernie.geckolib.core.animation.RawAnimation;
import software.bernie.geckolib.core.animation.AnimationState;
import software.bernie.geckolib.core.animation.AnimationController;
import software.bernie.geckolib.core.animation.AnimatableManager;
import software.bernie.geckolib.core.animatable.instance.AnimatableInstanceCache;
import software.bernie.geckolib.animatable.GeoEntity;

import net.minecraftforge.registries.ForgeRegistries;
import net.minecraftforge.network.PlayMessages;
import net.minecraftforge.network.NetworkHooks;

import net.minecraft.world.level.ServerLevelAccessor;
import net.minecraft.world.level.Level;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.entity.npc.Villager;
import net.minecraft.world.entity.monster.Monster;
import net.minecraft.world.entity.ai.goal.target.NearestAttackableTargetGoal;
import net.minecraft.world.entity.ai.goal.target.HurtByTargetGoal;
import net.minecraft.world.entity.ai.goal.RandomLookAroundGoal;
import net.minecraft.world.entity.ai.goal.MeleeAttackGoal;
import net.minecraft.world.entity.ai.goal.FloatGoal;
import net.minecraft.world.entity.ai.attributes.Attributes;
import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
import net.minecraft.world.entity.SpawnGroupData;
import net.minecraft.world.entity.Pose;
import net.minecraft.world.entity.MobType;
import net.minecraft.world.entity.MobSpawnType;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.EntityDimensions;
import net.minecraft.world.damagesource.DamageTypes;
import net.minecraft.world.damagesource.DamageSource;
import net.minecraft.world.DifficultyInstance;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.network.syncher.SynchedEntityData;
import net.minecraft.network.syncher.EntityDataSerializers;
import net.minecraft.network.syncher.EntityDataAccessor;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.Packet;
import net.minecraft.nbt.CompoundTag;

import net.mcreator.deaceased.procedures.ThediggerOnInitialEntitySpawnProcedure;
import net.mcreator.deaceased.procedures.ThediggerOnEntityTickUpdateProcedure;
import net.mcreator.deaceased.procedures.ThediggerEntityIsHurtProcedure;
import net.mcreator.deaceased.procedures.ThediggerDeathTimeIsReachedProcedure;
import net.mcreator.deaceased.init.DeaceasedModEntities;

import javax.annotation.Nullable;

public class ThediggerEntity extends Monster implements GeoEntity {
	public static final EntityDataAccessor<Boolean> SHOOT = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.BOOLEAN);
	public static final EntityDataAccessor<String> ANIMATION = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.STRING);
	public static final EntityDataAccessor<String> TEXTURE = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.STRING);
	public static final EntityDataAccessor<Integer> DATA_state = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.INT);
	public static final EntityDataAccessor<Integer> DATA_attackcooldown = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.INT);
	public static final EntityDataAccessor<Boolean> DATA_actionlock = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.BOOLEAN);
	public static final EntityDataAccessor<Integer> DATA_animationclock = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.INT);
	public static final EntityDataAccessor<Integer> DATA_safex = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.INT);
	public static final EntityDataAccessor<Integer> DATA_safey = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.INT);
	public static final EntityDataAccessor<Integer> DATA_safez = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.INT);
	public static final EntityDataAccessor<Boolean> DATA_under = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.BOOLEAN);
	public static final EntityDataAccessor<Boolean> DATA_safepo = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.BOOLEAN);
	public static final EntityDataAccessor<Boolean> DATA_assasin = SynchedEntityData.defineId(ThediggerEntity.class, EntityDataSerializers.BOOLEAN);
	private final AnimatableInstanceCache cache = GeckoLibUtil.createInstanceCache(this);
	private boolean swinging;
	private boolean lastloop;
	private long lastSwing;
	public String animationprocedure = "empty";

	public ThediggerEntity(PlayMessages.SpawnEntity packet, Level world) {
		this(DeaceasedModEntities.THEDIGGER.get(), world);
	}

	public ThediggerEntity(EntityType<ThediggerEntity> type, Level world) {
		super(type, world);
		xpReward = 0;
		setNoAi(false);
		setMaxUpStep(50f);
		noCulling = true;
	}

	@Override
	protected void defineSynchedData() {
		super.defineSynchedData();
		this.entityData.define(SHOOT, false);
		this.entityData.define(ANIMATION, "undefined");
		this.entityData.define(TEXTURE, "digger");
		this.entityData.define(DATA_state, 0);
		this.entityData.define(DATA_attackcooldown, 0);
		this.entityData.define(DATA_actionlock, false);
		this.entityData.define(DATA_animationclock, 0);
		this.entityData.define(DATA_safex, 0);
		this.entityData.define(DATA_safey, 0);
		this.entityData.define(DATA_safez, 0);
		this.entityData.define(DATA_under, false);
		this.entityData.define(DATA_safepo, false);
		this.entityData.define(DATA_assasin, false);
	}

	public void setTexture(String texture) {
		this.entityData.set(TEXTURE, texture);
	}

	public String getTexture() {
		return this.entityData.get(TEXTURE);
	}

	@Override
	public Packet<ClientGamePacketListener> getAddEntityPacket() {
		return NetworkHooks.getEntitySpawningPacket(this);
	}

	@Override
	protected void registerGoals() {
		super.registerGoals();
		this.goalSelector.addGoal(1, new MeleeAttackGoal(this, 1.2, false) {
			@Override
			protected double getAttackReachSqr(LivingEntity entity) {
				return this.mob.getBbWidth() * this.mob.getBbWidth() + entity.getBbWidth();
			}
		});
		this.targetSelector.addGoal(2, new HurtByTargetGoal(this));
		this.targetSelector.addGoal(3, new NearestAttackableTargetGoal(this, Player.class, false, false));
		this.targetSelector.addGoal(4, new NearestAttackableTargetGoal(this, Villager.class, false, false));
		this.goalSelector.addGoal(5, new RandomLookAroundGoal(this));
		this.goalSelector.addGoal(6, new FloatGoal(this));
	}

	@Override
	public MobType getMobType() {
		return MobType.UNDEAD;
	}

	@Override
	public SoundEvent getHurtSound(DamageSource ds) {
		return ForgeRegistries.SOUND_EVENTS.getValue(new ResourceLocation("entity.generic.hurt"));
	}

	@Override
	public SoundEvent getDeathSound() {
		return ForgeRegistries.SOUND_EVENTS.getValue(new ResourceLocation("entity.generic.death"));
	}

	@Override
	public boolean hurt(DamageSource source, float amount) {
		ThediggerEntityIsHurtProcedure.execute(this.level(), this.getX(), this.getY(), this.getZ(), this, source.getEntity());
		if (source.is(DamageTypes.FALL))
			return false;
		return super.hurt(source, amount);
	}

	@Override
	public SpawnGroupData finalizeSpawn(ServerLevelAccessor world, DifficultyInstance difficulty, MobSpawnType reason, @Nullable SpawnGroupData livingdata, @Nullable CompoundTag tag) {
		SpawnGroupData retval = super.finalizeSpawn(world, difficulty, reason, livingdata, tag);
		ThediggerOnInitialEntitySpawnProcedure.execute(this);
		return retval;
	}

	@Override
	public void addAdditionalSaveData(CompoundTag compound) {
		super.addAdditionalSaveData(compound);
		compound.putString("Texture", this.getTexture());
		compound.putInt("Datastate", this.entityData.get(DATA_state));
		compound.putInt("Dataattackcooldown", this.entityData.get(DATA_attackcooldown));
		compound.putBoolean("Dataactionlock", this.entityData.get(DATA_actionlock));
		compound.putInt("Dataanimationclock", this.entityData.get(DATA_animationclock));
		compound.putInt("Datasafex", this.entityData.get(DATA_safex));
		compound.putInt("Datasafey", this.entityData.get(DATA_safey));
		compound.putInt("Datasafez", this.entityData.get(DATA_safez));
		compound.putBoolean("Dataunder", this.entityData.get(DATA_under));
		compound.putBoolean("Datasafepo", this.entityData.get(DATA_safepo));
		compound.putBoolean("Dataassasin", this.entityData.get(DATA_assasin));
	}

	@Override
	public void readAdditionalSaveData(CompoundTag compound) {
		super.readAdditionalSaveData(compound);
		if (compound.contains("Texture"))
			this.setTexture(compound.getString("Texture"));
		if (compound.contains("Datastate"))
			this.entityData.set(DATA_state, compound.getInt("Datastate"));
		if (compound.contains("Dataattackcooldown"))
			this.entityData.set(DATA_attackcooldown, compound.getInt("Dataattackcooldown"));
		if (compound.contains("Dataactionlock"))
			this.entityData.set(DATA_actionlock, compound.getBoolean("Dataactionlock"));
		if (compound.contains("Dataanimationclock"))
			this.entityData.set(DATA_animationclock, compound.getInt("Dataanimationclock"));
		if (compound.contains("Datasafex"))
			this.entityData.set(DATA_safex, compound.getInt("Datasafex"));
		if (compound.contains("Datasafey"))
			this.entityData.set(DATA_safey, compound.getInt("Datasafey"));
		if (compound.contains("Datasafez"))
			this.entityData.set(DATA_safez, compound.getInt("Datasafez"));
		if (compound.contains("Dataunder"))
			this.entityData.set(DATA_under, compound.getBoolean("Dataunder"));
		if (compound.contains("Datasafepo"))
			this.entityData.set(DATA_safepo, compound.getBoolean("Datasafepo"));
		if (compound.contains("Dataassasin"))
			this.entityData.set(DATA_assasin, compound.getBoolean("Dataassasin"));
	}

	@Override
	public void baseTick() {
		super.baseTick();
		ThediggerOnEntityTickUpdateProcedure.execute(this.level(), this.getX(), this.getY(), this.getZ(), this);
		this.refreshDimensions();
	}

	@Override
	public EntityDimensions getDimensions(Pose p_33597_) {
		return super.getDimensions(p_33597_).scale((float) 1);
	}

	public static void init() {
	}

	public static AttributeSupplier.Builder createAttributes() {
		AttributeSupplier.Builder builder = Mob.createMobAttributes();
		builder = builder.add(Attributes.MOVEMENT_SPEED, 0.3);
		builder = builder.add(Attributes.MAX_HEALTH, 20);
		builder = builder.add(Attributes.ARMOR, 0);
		builder = builder.add(Attributes.ATTACK_DAMAGE, 1);
		builder = builder.add(Attributes.FOLLOW_RANGE, 64);
		return builder;
	}

	private PlayState movementPredicate(AnimationState event) {
		if (this.animationprocedure.equals("empty")) {
			return event.setAndContinue(RawAnimation.begin().thenLoop("idle"));
		}
		return PlayState.STOP;
	}

	String prevAnim = "empty";

	private PlayState procedurePredicate(AnimationState event) {
		if (!animationprocedure.equals("empty") && event.getController().getAnimationState() == AnimationController.State.STOPPED || (!this.animationprocedure.equals(prevAnim) && !this.animationprocedure.equals("empty"))) {
			if (!this.animationprocedure.equals(prevAnim))
				event.getController().forceAnimationReset();
			event.getController().setAnimation(RawAnimation.begin().thenPlay(this.animationprocedure));
			if (event.getController().getAnimationState() == AnimationController.State.STOPPED) {
				this.animationprocedure = "empty";
				event.getController().forceAnimationReset();
			}
		} else if (animationprocedure.equals("empty")) {
			prevAnim = "empty";
			return PlayState.STOP;
		}
		prevAnim = this.animationprocedure;
		return PlayState.CONTINUE;
	}

	@Override
	protected void tickDeath() {
		++this.deathTime;
		if (this.deathTime == 20) {
			this.remove(ThediggerEntity.RemovalReason.KILLED);
			this.dropExperience();
			ThediggerDeathTimeIsReachedProcedure.execute(this.level(), this.getX(), this.getY(), this.getZ());
		}
	}

	public String getSyncedAnimation() {
		return this.entityData.get(ANIMATION);
	}

	public void setAnimation(String animation) {
		this.entityData.set(ANIMATION, animation);
	}

	@Override
	public void registerControllers(AnimatableManager.ControllerRegistrar data) {
		data.add(new AnimationController<>(this, "movement", 4, this::movementPredicate));
		data.add(new AnimationController<>(this, "procedure", 4, this::procedurePredicate));
	}

	@Override
	public AnimatableInstanceCache getAnimatableInstanceCache() {
		return this.cache;
	}
}
